File: src/test/scala/tlschannel/helpers/SslContextFactory.scala
package tlschannel.helpers

import javax.net.ssl.SSLContext
import javax.net.ssl.KeyManagerFactory
import javax.net.ssl.TrustManagerFactory
import java.security.KeyStore

import scala.util.Using

class SslContextFactory(val protocol: String = "TLSv1.2") {

  val defaultContext = {
    val sslContext = SSLContext.getInstance(protocol)
    val ks = KeyStore.getInstance("JKS")
    Using.resource(getClass.getClassLoader.getResourceAsStream("keystore.jks")) { keystoreFile =>
      ks.load(keystoreFile, "password".toCharArray())
      val tmf = TrustManagerFactory.getInstance(TrustManagerFactory.getDefaultAlgorithm)
      tmf.init(ks)
      val kmf = KeyManagerFactory.getInstance(KeyManagerFactory.getDefaultAlgorithm)
      kmf.init(ks, "password".toCharArray())
      sslContext.init(kmf.getKeyManagers, tmf.getTrustManagers, null)
    }
    sslContext
  }

  val allCiphers = {
    ciphers(defaultContext).sorted
  }

  private def ciphers(ctx: SSLContext): Seq[String] = {
    ctx
      .createSSLEngine()
      .getSupportedCipherSuites
      .toSeq
      // this is not a real cipher, but a hack actually
      .filterNot(_ == "TLS_EMPTY_RENEGOTIATION_INFO_SCSV")
      // disable problematic ciphers
      .filterNot { c =>
        Seq(
          "TLS_KRB5_EXPORT_WITH_DES_CBC_40_MD5",
          "TLS_KRB5_EXPORT_WITH_DES_CBC_40_SHA",
          "TLS_KRB5_EXPORT_WITH_RC4_40_MD5",
          "TLS_KRB5_EXPORT_WITH_RC4_40_SHA",
          "TLS_KRB5_WITH_3DES_EDE_CBC_MD5",
          "TLS_KRB5_WITH_3DES_EDE_CBC_SHA",
          "TLS_KRB5_WITH_DES_CBC_MD5",
          "TLS_KRB5_WITH_DES_CBC_SHA",
          "TLS_KRB5_WITH_RC4_128_MD5",
          "TLS_KRB5_WITH_RC4_128_SHA",
          "SSL_RSA_EXPORT_WITH_DES40_CBC_SHA",
          "SSL_RSA_EXPORT_WITH_RC4_40_MD5"
        ).contains(c)
      }
      // No SHA-2 with TLS < 1.2
      .filterNot { c =>
        !Set("TLSv1.2", "TLSv1.3").contains(protocol) && (c.endsWith("_SHA256") || c.endsWith("_SHA384"))
      }
      // Disable cipher only supported in TLS >= 1.3
      .filterNot { c =>
        protocol < "TLSv1.3" &&
        Set(
          "TLS_AES_128_GCM_SHA256",
          "TLS_AES_256_GCM_SHA384"
        ).contains(c)
      }
      // https://bugs.openjdk.java.net/browse/JDK-8224997
      .filterNot(_.endsWith("_CHACHA20_POLY1305_SHA256"))
      // Anonymous ciphers are problematic because the are disabled in some VMs
      .filterNot(_.contains("_anon_"))
  }

}

object SslContextFactory {

  val tlsMaxDataSize = math.pow(2, 14).toInt
  val certificateCommonName = "name" // must match what's in the certificates

}

File: src/test/scala/tlschannel/helpers/SslContextFactory.java
package tlschannel.helpers;

import java.io.IOException;
import java.io.InputStream;
import java.security.*;
import java.util.Arrays;
import java.util.List;
import java.util.Objects;
import java.util.stream.Collectors;
import javax.net.ssl.KeyManagerFactory;
import javax.net.ssl.SSLContext;
import javax.net.ssl.TrustManagerFactory;

public class SslContextFactory {

    public static final int tlsMaxDataSize = (int) Math.pow(2, 14);
    public static final String certificateCommonName = "name"; // must match what's in the certificates

    private final String protocol;
    private final SSLContext defaultContext;
    private final List<String> allCiphers;

    public SslContextFactory(String protocol) {
        this.protocol = protocol;
        try {
            SSLContext sslContext = SSLContext.getInstance(protocol);
            KeyStore ks = KeyStore.getInstance("JKS");
            try (InputStream keystoreFile = getClass().getClassLoader().getResourceAsStream("keystore.jks")) {
                ks.load(keystoreFile, "password".toCharArray());
                TrustManagerFactory tmf = TrustManagerFactory.getInstance(TrustManagerFactory.getDefaultAlgorithm());
                tmf.init(ks);
                KeyManagerFactory kmf = KeyManagerFactory.getInstance(KeyManagerFactory.getDefaultAlgorithm());
                kmf.init(ks, "password".toCharArray());
                sslContext.init(kmf.getKeyManagers(), tmf.getTrustManagers(), null);
            }
            this.defaultContext = sslContext;
        } catch (IOException | GeneralSecurityException e) {
            throw new RuntimeException(e);
        }
        this.allCiphers = ciphers(defaultContext).stream().sorted().collect(Collectors.toList());
    }

    public SslContextFactory() {
        this("TLSv1.2");
    }

    private List<String> ciphers(SSLContext ctx) {
        return Arrays.stream(ctx.createSSLEngine().getSupportedCipherSuites())
                // this is not a real cipher, but a hack actually
                .filter(c -> !Objects.equals(c, "TLS_EMPTY_RENEGOTIATION_INFO_SCSV"))
                // disable problematic ciphers
                .filter(c -> !Arrays.asList(
                                "TLS_KRB5_EXPORT_WITH_DES_CBC_40_MD5",
                                "TLS_KRB5_EXPORT_WITH_DES_CBC_40_SHA",
                                "TLS_KRB5_EXPORT_WITH_RC4_40_MD5",
                                "TLS_KRB5_EXPORT_WITH_RC4_40_SHA",
                                "TLS_KRB5_WITH_3DES_EDE_CBC_MD5",
                                "TLS_KRB5_WITH_3DES_EDE_CBC_SHA",
                                "TLS_KRB5_WITH_DES_CBC_MD5",
                                "TLS_KRB5_WITH_DES_CBC_SHA",
                                "TLS_KRB5_WITH_RC4_128_MD5",
                                "TLS_KRB5_WITH_RC4_128_SHA",
                                "SSL_RSA_EXPORT_WITH_DES40_CBC_SHA",
                                "SSL_RSA_EXPORT_WITH_RC4_40_MD5")
                        .contains(c))
                // No SHA-2 with TLS < 1.2
                .filter(c -> Arrays.asList("TLSv1.2", "TLSv1.3").contains(protocol)
                        || !c.endsWith("_SHA256") && !c.endsWith("_SHA384"))
                // Disable cipher only supported in TLS >= 1.3
                .filter(c -> protocol.compareTo("TLSv1.3") > 0
                        || !Arrays.asList("TLS_AES_128_GCM_SHA256", "TLS_AES_256_GCM_SHA384")
                                .contains(c))
                // https://bugs.openjdk.java.net/browse/JDK-8224997
                .filter(c -> !c.endsWith("_CHACHA20_POLY1305_SHA256"))
                // Anonymous ciphers are problematic because they are disabled in some VMs
                .filter(c -> !c.contains("_anon_"))
                .collect(Collectors.toList());
    }

    public SSLContext defaultContext() {
        return defaultContext;
    }

    public List<String> getAllCiphers() {
        return allCiphers;
    }
}
